tests/test_ingest.py

```diff
@@ -25,38 +25,336 @@
+        new_run = self.outputRun + "_bad_exposure"
+        with self.assertRaises(ConflictingDefinitionError):
+            self.task.run(files, run=new_run)
+
+    def testBadFile(self):
+        """Try to ingest a bad file."""
+        files = [self.bad_metadata_file]
+
+        with self.assertRaises(RuntimeError) as cm:
+            # Default is to raise an error at the end
+            self.task.run(files, run=self.outputRun)
+        self.assertIn("Some failures", str(cm.exception))
+
+        # Including a good file will result in ingest working but still
+        # raises (we might want to move this to solely happen in the
+        # command line invocation).
+        files.append(self.good_file)
+
+        # Also include a file with unkown instrument
+        files.append(self.bad_instrument_file)
+
+        with self.assertRaises(RuntimeError):
+            self.task.run(files, run=self.outputRun)
+        datasets = list(self.butler.registry.queryDatasets("raw_dict", collections=self.outputRun))
+        self.assertEqual(len(datasets), 1)
+
+        # Fail fast will trigger a run time error with different text.
+        # Use a different output run to be sure we are not failing because
+        # of the attempt to ingest twice.
+        self.task.config.failFast = True
+        new_run = self.outputRun + "b"
+        with self.assertRaises(RuntimeError) as cm:
+            self.task.run([self.bad_metadata_file, self.good_file], run=new_run)
+        self.assertIn("Problem extracting metadata", str(cm.exception))
+
+        # Attempt to ingest good file again -- this will fail for a different
+        # reason than failed metadata extraction
+        with self.assertRaises(ConflictingDefinitionError):
+            self.task.run([self.good_file], run=self.outputRun)
+
+        # Ingest a file with good metadata but unknown instrument
+        with self.assertRaises(RuntimeError) as cm:
+            self.task.run([self.bad_instrument_file], run=self.outputRun)
+        self.assertIn("Instrument HSC", str(cm.exception))
+
+        # Ingest of a metadata index file that will fail translation
+        with self.assertRaises(RuntimeError) as cm:
+            self.task.run([os.path.join(INGESTDIR, "indexed_data", "metadata_subdir", "_index.json")])
+        self.assertIn("Problem extracting metadata", str(cm.exception))
+
+        # ingest of a bad index file
+        with self.assertRaises(RuntimeError) as cm:
+            self.task.run([os.path.join(INGESTDIR, "indexed_data", "bad_index", "_index.json")])
+        self.assertIn("Problem reading index file", str(cm.exception))
+
+        # ingest of an implied bad index file
+        with self.assertRaises(RuntimeError) as cm:
+            self.task.run([os.path.join(INGESTDIR, "indexed_data", "bad_implied", "dataset_2.yaml")])
+
+    def testCallbacks(self):
+        """Test the callbacks for failures."""
```

Comment:
I feel like this block of tests might be clearer if they were all separate `testX` methods (maybe with a `setup_callbacks()` to prep things), instead of trying to clear the appropriate callback lists each time. I'll leave it up to you whether that's worth changing.